### PR TITLE
faster per_token_group_quant_8bit kernel with stride support

### PR DESCRIPTION
**Pull request by strgrb**
## Motivation

I want some new features for per_token_group_quant_8bit for future optimize:
- support scale.dtype=float even if ue8m0 is on, to test ue8m0 quant on hopper gpus
- support stride for future fusion
So I decided to rewrite it with cute, and got some benefits.

Following test is tested on H20:
```
per-token-group-quant-8bit-performance:
   batch_size  seq_len  group_size            dst_dtype  SGL Kernel  new kernel
0           1     1024         128           torch.int8   36.800001    36.063999
1           1     1024         128  torch.float8_e4m3fn   37.983999    32.992002
2           1     4096         128           torch.int8  117.183998   117.311999
3           1     4096         128  torch.float8_e4m3fn  128.191993   113.407999
4           1     8192         128           torch.int8  223.488003   223.391995
5           1     8192         128  torch.float8_e4m3fn  244.448006   218.367994
6           1    16384         128           torch.int8  435.167998   435.615987
7           1    16384         128  torch.float8_e4m3fn  477.600008   426.815987
```

and in this case, the result differs, but I think my impl is correct:
```
import sgl_kernel
import torch
from sgl_kernel.gemm import sgl_per_token_group_quant_8bit as per_token_group_quant_with_prologue
from sgl_kernel import sgl_per_token_group_quant_fp8
import torch

a = torch.randn([3, 128], device='cuda', dtype=torch.bfloat16)
group_size = 128

def allocate_qs(input, q_dtype, aligned=False, scale_ue8m0=False):
  q = torch.empty_like(input, device='cuda', dtype=q_dtype)
  m, n = input.shape
  if scale_ue8m0:
    am = (m + 3) // 4 * 4
    an = (n // group_size + 3) // 4 * 4
    s = torch.zeros([an // 4, am], device='cuda', dtype=torch.int).T[:m, :]
  elif aligned:
    am = (m + 3) // 4 * 4
    s = torch.empty([n // group_size, am], device='cuda', dtype=torch.float32).T[:m, :]
  else:
    s = torch.empty([m, n // group_size], device='cuda', dtype=torch.float32)
  return q, s

def my_quant(q_dtype, aligned, scale_ue8m0):
  q, s = allocate_qs(a, q_dtype, aligned, scale_ue8m0)
  per_token_group_quant_with_prologue(a, q, s, group_size, 1e-10, -448, 448, scale_ue8m0)
  return q, s

def sgl_quant(q_dtype, aligned, scale_ue8m0):
  q, s = allocate_qs(a, q_dtype, aligned, scale_ue8m0)
  sgl_per_token_group_quant_fp8(a, q, s, group_size, 1e-10, -448, 448, scale_ue8m0=scale_ue8m0)
  return q, s

q_dtype = torch.int8
aligned = True
scale_ue8m0 = True
q1, s1 = my_quant(q_dtype, aligned, scale_ue8m0)
q2, s2 = sgl_quant(q_dtype, aligned, scale_ue8m0)
```

If `hidden_dim // group_size` cannot be divided by 4, the row_size calculated here is not correct. 
https://github.com/sgl-project/sglang/blob/8abd3e77feca9ed740356c1b879e524d09482fb2/sgl-kernel/csrc/gemm/per_token_group_quant_8bit.cu#L56

## Modifications

A new kernel written with cute.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: sgl-kernel/csrc/gemm/per_token_group_quant_8bit_prologue.cu [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+
+PerTokenGroupQuantDispatcher dispatcher;
+
+}  // namespace
+
+void check_contiguous_except_last(torch::Tensor x) {
+  CHECK(x.stride(x.dim() - 1) == 1);
+  for (int i = 0; i < x.dim() - 3; ++i) {
+    CHECK(x.stride(i) / x.stride(i + 1) == x.size(i + 1));
+  }
```
[the remark this answers is not in the document]
Comment (strgrb) [reply sgl-kernel/csrc/gemm/per_token_group_quant_8bit_prologue.cu:379]: if dim[i] is contiguous, stride[i-1] / stride[i] == dim[i],  I set x.size(i+1), so last check is x.size(x.dim() - 2), that is dim -2, and last dim should not be checked.

**Comment by BBuf** [comment]
@strgrb  You're right, this cute implementation is really cool! Are you planning to help us fix the bug in the original kernel where `hidden_dim // group_size` isn't divisible by 4?

**Comment by BBuf** [comment]
@strgrb By the way, where does this cute implementation get its main speedup compared to the original, and do you have any NCU analysis results or similar? Thanks.

**Comment by HydraQYH** [comment]
> @strgrb You're right, this cute implementation is really cool! Are you planning to help us fix the bug in the original kernel where `hidden_dim // group_size` isn't divisible by 4?

https://github.com/sgl-project/sglang/blob/0c8dab9e67b1fe0d274a27af03540b2ce5525a37/sgl-kernel/csrc/gemm/per_token_group_quant_8bit.cu#L55
@strgrb @BBuf Should there be a cdiv here?

**Comment by strgrb** [comment]
> @strgrb By the way, where does this cute implementation get its main speedup compared to the original, and do you have any NCU analysis results or similar? Thanks.

In my kernel, I avoid reload input again, but I see L2 cache hit, so it's not the problem since it's a compute-intensive kernel. By checking the SASS code in ncu profile, I finally found the cause, the original kernel do not use fp8 instrinct to convert . Following is sass code from my kernel.
image

I guess the reason maybe the original kernel use `c10::Float8_e4m3` , not `__nv_fp8_e4m3`?

**Comment by strgrb** [comment]
> @strgrb You're right, this cute implementation is really cool! Are you planning to help us fix the bug in the original kernel where `hidden_dim // group_size` isn't divisible by 4?

I have some other plan this week, I'll take a look at next week.

**Comment by HydraQYH** [comment]
> > @strgrb You're right, this cute implementation is really cool! Are you planning to help us fix the bug in the original kernel where `hidden_dim // group_size` isn't divisible by 4?
> 
> I have some other plan this week, I'll take a look at next week.

@strgrb Hi, I want to ask a simple question because I have never read the original kernel code. What will happen when `hidden_dim // group_size` isn't divisible by 4?
I'm guessing it's because of some additional alignment requirement, but I'm not sure.

**Comment by strgrb** [comment]
> > > @strgrb You're right, this cute implementation is really cool! Are you planning to help us fix the bug in the original kernel where `hidden_dim // group_size` isn't divisible by 4?
> > 
> > 
> > I have some other plan this week, I'll take a look at next week.
> 
> @strgrb Hi, I want to ask a simple question because I have never read the original kernel code. What will happen when `hidden_dim // group_size` isn't divisible by 4? I'm guessing it's because of some additional alignment requirement, but I'm not sure.

@HydraQYH According to `fp8_kernel.py` , both token_count dim and hidden dim are aligned to 16B if ue8m0 is on. Since token_count dim is contiguous, it should be aligned, and ue8m0 scale is packed to int32 with hidden dim, so hidden dim should also be aligned.  The unpacked col_idx should be calculated by real `hidden_dim // group_size` .

File: sgl-kernel/csrc/gemm/per_token_group_quant_8bit_prologue.cu [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+  auto in_stride = make_stride(in_hidden_stride, _1{});
+  auto out_stride = make_stride(out_hidden_stride, _1{});
+
+  int num_threads = 128;
+  dim3 grid(token_count, (hidden_dim + VEC_SIZE * num_threads - 1) / (VEC_SIZE * num_threads));
+  auto warp_tiler = make_shape(Int<VEC_SIZE * 32>{});
+  if (scale_ue8m0 && std::is_same<TS, uint8_t>::value) {
+    auto scale_layout = make_layout(
+        make_shape(token_count, make_shape(_4{}, ceil_div(hidden_dim / GROUP_SIZE, 4))),
+        make_stride(_4{}, make_stride(scale_stride0, 4 * scale_stride1)));
```
Comment (yue-zhang-2025) [inline sgl-kernel/csrc/gemm/per_token_group_quant_8bit_prologue.cu:172]: Hi @strgrb, the work looks amazing! I'm trying to understand the make_stride here - iiuc the first _4{} represent the number of bytes take by each token (stride for token dimension), wondering are we assuming we have <= 4 groups per token? if we have > 4 groups per token the _4{} would not be enough? maybe I have wrong understanding

Comment (strgrb) [reply sgl-kernel/csrc/gemm/per_token_group_quant_8bit_prologue.cu:172]: The scale original shape is [token_count, hidden_dim // group_size], and in scale_ue8m0 case, both dims are padded to multiple of 4 for alignment requirement of tma, and this `_4` is `sizeof(int) / sizeof(uint8_t)`, the pack size of ue8m0
